README
Tuned: Daemon for monitoring and adaptive tuning of system devices.

(This is tuned 2.0 with a new code base. If you are looking for the older
version, please check out branch '1.0' in our Git repository.)

How to use it
-------------

In Fedora, Red Hat Enterprise Linux, and their derivates install tuned package
(optionally tuned-utils, tuned-utils-systemtap, and tuned-profiles-compat):

# yum install tuned

After the installation, start the tuned service:

# systemctl start tuned

You might also want to run tuned whenever your machine starts:

# systemctl enable tuned

If the daemon is running you can easily control it using 'tuned-adm' command
line utility. This tool communicates with the daemon over DBus. Any user can
list the available profiles and see which one is active. But the profiles can
be switched only by root user or by any user with physical console allocated
on the machine (X11, physical tty, but no SSH).

To see the current active profile, run:

$ tuned-adm active

To list all available profiles, run:

$ tuned-adm list

To switch to a different profile, run:

# tuned-adm profile <profile-name>

And to disable all tunings, run:

# tuned-adm recommend
Recommend profile suitable for your system. Currently only static detection is
implemented - it decides according to data in /etc/system-release-cpe and
virt-what output. The rules for autodetection are defined in recommend.conf
in profile directory. They can be overriden by user by putting the
recommend.conf into /etc/tuned. The default rules recommends profiles targeted
to the best performance or balanced profile if unsure.

# tuned-adm off

Your profile choice is also written into /etc/tuned/active_profile and this
choice is used when the daemon is restarted (e.g. with the machine reboot).


Available tunings
-----------------

We are currenlty working on many new tuning features. Some are described in
the manual pages, some are yet undocumented.


Authors
-------

The best way to contact the authors of the project is to use our mailing list:
power-management@lists.fedoraproject.org

In case you want to contact individual author, you will find the e-mail
address in every commit message in our Git repository:
http://git.fedorahosted.org/cgit/tuned.git

You can also join #fedora-power IRC channel on Freenode.


License
-------

Copyright (C) 2008-2012 Red Hat, Inc.

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.

Full text of the license is enclosed in COPYING file.


ChangeLog
[33mcommit adf6d855789f1e5738ae0a113a96d914b480e23f[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Tue Jan 29 16:05:50 2013 +0100

    udev rules: ask tuned to restart ktune instead of restarting it directly

[33mcommit afd26f62774daf81a9f95d256d415547300f7757[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Tue Jan 29 13:26:25 2013 +0100

    tuned: restart ktune 10 seconds after last SIGUSR2 was received
    
    When multiple devices are added into the system, udev restarted ktune
    daemon for each new device. This can cause races within ktune. This
    mechanism allows triggering just one ktune restart when multiple
    devices are inserted.

[33mcommit 6fc759a9fff5801a9f80e672207fc41f95b3ae7b[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Tue Jan 29 14:32:30 2013 +0100

    functions: implement support for readahead restore

[33mcommit f46c9d05477f93f31e3c5ee1c24637f4adde743c[m
Author: Jaroslav Škarvada <jskarvad@redhat.com>
Date:   Thu Jan 3 17:03:43 2013 +0100

    profiles: removed the nobarrier from the virtual-guest profile
    
    resolves: #886956

[33mcommit 355fd87b1b2304863fe04266e290dc99394587ad[m
Author: Jaroslav Škarvada <jskarvad@redhat.com>
Date:   Thu Jan 3 17:01:48 2013 +0100

    disabled transparent hugepages in latency-performance profile
    
    resolves: rhbz#887355

[33mcommit 66d8e396db3845463abc181c541ee09c686b1ff1[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Fri Dec 7 14:06:09 2012 +0100

    {disk,net}devstat: fix typo in usage

[33mcommit c1b7f35546a23b409ca0e1479d928b8d7f57c06b[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Wed Oct 10 10:47:26 2012 +0200

    udev: add rule to restart ktune in new block device is added
    
    Resolves: #847445

[33mcommit 64b45e8a04b06c0f554c64b458c590c9904f7f5f[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Thu Sep 27 13:40:38 2012 +0200

    profiles: do not disable barriers on devices with write back cache
    
    Resolves: #801561

[33mcommit 9e8f67009568adce6b9a2f1d4c4c828b2192cba8[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Mon Sep 24 15:08:19 2012 +0200

    tuned: create PID file with proper permissions
    
    Resolves: #845336

[33mcommit 84536e5a5917282821a654cb5d47ac3505d9506e[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Mon Sep 24 14:51:59 2012 +0200

    ktune: validate value of ELEVATOR_TUNE_DEVS
    
    Resolves: #784308

[33mcommit d4c0b5a2d145da188b995419764f9c969b9afdbb[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Tue Mar 20 14:29:04 2012 +0100

    update systemd unit file install path
    
    required by /lib -> /usr/lib feature

[33mcommit 9dfa60a1f4166aab80e400d88a3ee28bcc256efc[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Tue Mar 20 14:16:42 2012 +0100

    spec: missing system-units build requirement

[33mcommit cbb2bf6e85117a0f245de5010e70c75a8ccc35b6[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Tue Mar 20 13:55:48 2012 +0100

    version bump

[33mcommit 0427cea0d801b03ab5fbfdeb3c26922c6eaa2247[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Mon Mar 5 08:51:53 2012 +0100

    netdevstat: cmdline error handling

[33mcommit 81590cf7705e982c7880c69d01ea6d02a6e7a2e4[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Mon Mar 5 08:45:50 2012 +0100

    diskdevstat: cmdline error handling

[33mcommit 3b10ac690bc6164b721236702cb5eb8493278f1e[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Fri Mar 2 09:59:29 2012 +0100

    virtual-guest: cleanup ktune.sh
    
    References: rhbz#740976, rhbz#740977

[33mcommit 90afd92add2facb02596a7e8e1341f252ae3684d[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Fri Mar 2 09:00:26 2012 +0100

    virtual-host: tune virtual memory settings
    
    References: rhbz#740976

[33mcommit 78012939d465405587a6607838f15e8c06cbd5a3[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Wed Jan 11 13:38:28 2012 +0100

    tuned-adm: sort list of available profiles

[33mcommit e30a77ede22df20232547b6243107b57ed0ad0ee[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Wed Jan 11 13:35:45 2012 +0100

    tuned-adm man: add virtual-host profile decription

[33mcommit 5984ad2de7e2e5ee76d54191f03515ae55739a4e[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Wed Jan 11 13:26:31 2012 +0100

    add virtual-host profile

[33mcommit 0e4826f78b14f219c876bc9d3b35834cd7a18311[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Wed Jan 11 12:59:48 2012 +0100

    tuned-adm: fix crash when active-profile file is missing

[33mcommit 0754d462ff72ed69b704b3a4a4d7a5ecc77df655[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Wed Jan 11 12:55:22 2012 +0100

    tuned-adm man: add virtual-guest profile decription

[33mcommit 8bb2e57c523cc27915b422d077ba5c644bd85f6a[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Wed Jan 11 12:48:19 2012 +0100

    add virtual-guest profile

[33mcommit 8bd483a47c4e00cf106fb16d50ac5d0c61f21ab8[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Wed Jan 11 12:40:23 2012 +0100

    functions: support for disk scheduler quantum setting

[33mcommit b5297bcc8cdb71b7c7fff17b7a18ea194952cc1c[m
Author: Jan Vcelak <jvcelak@redhat.com>
Date:   Wed Jan 11 12:39:47 2012 +0100

    functions: update disks detection mechanism

[33mcommit f11bae54bc26e2969ab1413995baeae472e75b60[m
Author: Jaroslav Škarvada <jskarvad@redhat.com>
Date:   Sat Dec 3 15:31:29 2011 +0100

    set_radeon_powersave: fixed profile name

[33mcommit 229b5f7874c406f702b7f1fd9d4e18a1df0e575e[m
Author: Jaroslav Škarvada <jskarvad@redhat.com>
Date:   Sat Dec 3 15:14:06 2011 +0100

    Changed Radeon PM to auto in laptop-battery-powersave
    
    It should be safer for presenters, because the more
    aggresive dynpm doesn't support multi heads.

[33mcommit 60f78e3dc83907d558913473acc145ce864d4c26[m
Author: Jaroslav Škarvada <jskarvad@redhat.com>
Date:   Sat Dec 3 14:32:39 2011 +0100

    Improved radeon powersavings settings in tuned profiles

[33mcommit 9280e4ee158792973112ea78c3848bc25609c51e[m

tuned_adm.py
# Copyright (C) 2008, 2009 Red Hat, Inc.
# Authors: Phil Knirsch
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
#

import sys
import os
import os.path
import glob
import platform

class Tuned_adm:

	def __init__(self, profile_dir = "/etc/tune-profiles"):
		self.profile_dir = os.path.normpath(profile_dir)
		self.active_file = os.path.join(profile_dir, "active-profile")
		self.arch = platform.machine()

	def error(self, msg, exit_code = 1):
		print >>sys.stderr, msg
		sys.exit(exit_code)

	def check_permissions(self):
		if not os.getuid() == 0:
			self.error("Only root can run this script.", 2)

	def run(self, args):
		if args[0] == "list":
			self.list()
		elif args[0] == "active":
			self.active()
			self.service_status("tuned")
			self.service_status("ktune")
		elif args[0] == "off":
			self.check_permissions()
			self.off()
		elif args[0] == "profile":
			if len(args) == 2:
				self.check_permissions()
				self.profile(args[1])
			else:
				self.error("Invalid profile specification. Use 'tuned-adm list' to get all available profiles.")
		else:
			self.error("Nonexistent argument '%s'." % args[0])

	def list(self):
		modes = sorted(os.listdir(self.profile_dir))
		if len(modes) > 0:
			print "Available profiles:"
			for mode in modes:
				dir = os.path.join(self.profile_dir, mode)
				if not os.path.isdir(dir):
					continue
				print "- %s" % mode

			self.active()
		else:
			print "No profiles defined."

	def active(self):
		print "Current active profile: %s" % self.get_active()

	def service_status(self, service):
		(enabled, running) = self.get_service_status(service)

		print "Service %s: %s, %s" % (service,
				"enabled" if enabled else "disabled",
				"running" if running else "stopped")

	def get_active(self):
		try:
			file = open(self.active_file, "r")
			profile = file.read()
			file.close()
		except IOError:
			profile = "off"

		return profile

	def set_active(self, profile):
		file = open(self.active_file, "w")
		file.write(profile)
		file.close()

	def get_service_status(self, service):
		enabled = os.system("chkconfig %s &>/dev/null" % service) == 0
		running = os.system("service %s status &>/dev/null" % service) == 0
		return (enabled, running)

	def verify_profile(self, profile):
		path = os.path.abspath(os.path.join(self.profile_dir, profile))

		if not path.startswith("%s/" % self.profile_dir):
			return False

		if not os.path.isdir(path):
			return False

		return True

	def remove(self, wildcard, filter = None):
		if filter == None:
			filter = lambda file: True

		files = glob.glob(wildcard)
		for f in files:
			if filter(f):
				os.unlink(f)

	def pick_config(self, name, profile_root):
		file = os.path.join(profile_root, name)
		(path, ext) = os.path.splitext(file)
		arch_specific = "%s.%s%s" % (path, self.arch, ext)
		if os.path.exists(arch_specific):
			return arch_specific
		elif os.path.exists(file):
			return file
		else:
			return None

	def off(self):
		self.set_active("off")

		# disable services
		os.system('service ktune stop')
		os.system('service tuned stop')
		os.system('chkconfig ktune off')
		os.system('chkconfig tuned off')

		# remove profile settings
		self.remove("/etc/ktune.d/*.conf", os.path.islink)
		self.remove("/etc/ktune.d/*.sh", os.path.islink)

		# restore previous ktune settings (if present)
		if os.path.exists("/etc/sysconfig/ktune.bckp"):
			os.rename("/etc/sysconfig/ktune.bckp", "/etc/sysconfig/ktune")
		if os.path.exists("/etc/tuned.conf.bckp") and os.path.exists("/etc/tuned.conf"):
			os.rename("/etc/tuned.conf.bckp", "/etc/tuned.conf")

	def profile(self, profile):
		enablektune = False
		enabletuned = False

		if not self.verify_profile(profile):
			self.error("Invalid profile. Use 'tuned-adm list' to get all available profiles.")

		profile_root = os.path.join(self.profile_dir, profile)

		# disabling services

		os.system('service ktune stop')
		os.system('service tuned stop')
		os.system('chkconfig ktune on')
		os.system('chkconfig tuned on')

		print >>sys.stderr, "Switching to profile '%s'" % profile
		self.set_active(profile)

		# ktune settings

		self.remove('/etc/ktune.d/tunedadm.sh')
		self.remove('/etc/ktune.d/tunedadm.conf')

		file = self.pick_config("ktune.sysconfig", profile_root)
		if file:
			enablektune = True
			os.rename('/etc/sysconfig/ktune', '/etc/sysconfig/ktune.bckp')
			os.symlink(file, '/etc/sysconfig/ktune')

		file = self.pick_config("ktune.sh", profile_root)
		if file:
			os.symlink(file, '/etc/ktune.d/tunedadm.sh')

		file = self.pick_config("sysctl.ktune", profile_root)
		if file:
			os.symlink(file, '/etc/ktune.d/tunedadm.conf')

		# tuned settings

		file = self.pick_config("tuned.conf", profile_root)
		if file:
			enabletuned = True
			os.rename('/etc/tuned.conf', '/etc/tuned.conf.bckp')
			os.symlink(file, '/etc/tuned.conf')

		# enabling services

		if enablektune:
			os.system('service ktune start')
			os.system('chkconfig ktune on')

		if enabletuned:
			os.system('service tuned start')
			os.system('chkconfig tuned on')

tuned_adm = Tuned_adm()
